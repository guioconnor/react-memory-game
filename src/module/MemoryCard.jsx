import React from 'react';
import styled from 'styled-components';

export const CARD_STATE = {
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  FOUND: 'FOUND',
};

const Card = styled.button`
  cursor: pointer;
  flex 1 0 0%;
  background: rgba(255, 255, 255, .8);
  opacity: ${state => (state.state === CARD_STATE.FOUND ? '.8' : '1')};
  color: #333;
  padding: 10%;
  margin: 10px;
  border-radius: 10px;
  box-shadow: 3px 3px 3px rgba(0, 0, 0, 0.2);
  position: relative;
`;

const CardContent = styled.div`
  display: ${state => (state.state === CARD_STATE.CLOSED ? 'none' : 'flex')};
  justify-content: center;
  align-items: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

const MemoryCard = ({ children, state, onClick }) => {
  return (
    <Card onClick={onClick} state={state}>
      <CardContent state={state}>{children}</CardContent>
    </Card>
  );
};

export default MemoryCard;
