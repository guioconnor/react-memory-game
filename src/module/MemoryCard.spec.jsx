import React, { PureComponent } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MemoryCard, { CARD_STATE } from './MemoryCard';

describe('<MemoryCard />', () => {
  const MockChildren = () => <div>Mock Children</div>;

  describe('when card state is CLOSED', () => {
    const container = shallow(
      <MemoryCard state={CARD_STATE.CLOSED}>
        <MockChildren />
      </MemoryCard>,
    );
    it('should render a closed button with children inside', () => {
      expect(toJson(container)).toMatchSnapshot('ClosedMemoryCard');
    });
  });

  describe('when card sate is OPEN', () => {
    const container = shallow(
      <MemoryCard state={CARD_STATE.OPEN}>
        <MockChildren />
      </MemoryCard>,
    );
    it('should render an open button with children inside', () => {
      expect(toJson(container)).toMatchSnapshot('OpenMemoryCard');
    });
  });

  describe('when card sate is FOUND', () => {
    const container = shallow(
      <MemoryCard state={CARD_STATE.FOUND}>
        <MockChildren />
      </MemoryCard>,
    );
    it('should render an found button with children inside', () => {
      expect(toJson(container)).toMatchSnapshot('FoundMemoryCard');
    });
  });

  describe('when card is clicked', () => {
    const mockClickHandler = jest.fn();
    const container = shallow(
      <MemoryCard state={CARD_STATE.FOUND} onClick={mockClickHandler}>
        <MockChildren />
      </MemoryCard>,
    );

    it('should call the callback fuction', () => {
      container.find('button').simulate('click');
      expect(mockClickHandler.mock.calls.length).toEqual(1);
    });
  });
});
