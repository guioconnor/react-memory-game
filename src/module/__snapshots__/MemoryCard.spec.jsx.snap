// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClosedMemoryCard 1`] = `
<button
  className="memory-game--memory-card memory-game--memory-card__closed"
>
  <CardContent>
    <MockChildren />
  </CardContent>
</button>
`;

exports[`FoundMemoryCard 1`] = `
<button
  className="memory-game--memory-card memory-game--memory-card__found"
>
  <CardContent>
    <MockChildren />
  </CardContent>
</button>
`;

exports[`OpenMemoryCard 1`] = `
<button
  className="memory-game--memory-card memory-game--memory-card__open"
>
  <CardContent>
    <MockChildren />
  </CardContent>
</button>
`;
